src/repositories/gitlab_updater.rs

```diff
@@ -0,0 +1,323 @@
+        for host in Self::hosts() {
+            let needs_update = conn
+                .query(
+                    "SELECT host_id
+                     FROM repositories
+                     WHERE host = $1 AND updated_at < NOW() - INTERVAL '1 day';",
+                    &[host],
+                )?
+                .into_iter()
+                .map(|row| row.get(0))
+                .collect::<Vec<String>>();
+
+            for chunk in needs_update.chunks(UPDATE_CHUNK_SIZE) {
+                if let Err(err) = self.update_repositories(host, &mut conn, &chunk) {
+                    if err.downcast_ref::<RateLimitReached>().is_some() {
+                        warn!("rate limit reached, blocked the Gitlab repository stats updater");
+                        return Ok(());
+                    }
+                    return Err(err);
+                }
+            }
+
+            updated += needs_update.len();
+        }
+
+        if updated == 0 {
+            info!("no Gitlab repository stats needed to be updated");
+        } else {
+            info!("finished updating Gitlab repository stats");
+        }
+        Ok(())
+    }
+
+    fn name() -> &'static str {
+        "Gitlab"
+    }
+
+    fn hosts() -> &'static [&'static str] {
+        &["gitlab.com", "gitlab.freedesktop.org"]
+    }
+
+    fn pool(&self) -> &Pool {
+        &self.pool
+    }
+}
+
+impl GitlabUpdater {
+    fn update_repositories(
+        &self,
+        host: &str,
+        conn: &mut Client,
+        node_ids: &[String],
+    ) -> Result<()> {
+        let response: GraphResponse<GraphProjects<Option<GraphProject>>> = self.graphql(
+            host,
+            GRAPHQL_UPDATE,
+            serde_json::json!({
+                "ids": node_ids,
+            }),
+        )?;
+
+        // The error is returned *before* we reach the rate limit, to ensure we always have an
+        // amount of API calls we can make at any time.
+        if let Some(data) = response.data {
+            // When a node is missing (for example if the repository was deleted or made private) the
+            // GraphQL API will return *both* a `null` instead of the data in the nodes list and a
+            // `NOT_FOUND` error in the errors list.
+            for node in &data.projects.nodes {
+                if let Some(node) = node {
+                    self.store_repository(
+                        conn,
+                        host,
+                        &node.id,
+                        &node.name_with_namespace,
+                        &node.description,
+                        &node.last_activity_at,
+                        node.star_count,
+                        node.forks_count,
+                        node.open_issues_count.unwrap_or(0),
+                    )?;
+                }
+            }
+            if !response.errors.is_empty() {
+                failure::bail!("error updating repositories: {:?}", response.errors);
```

Comment:
to match behaviour with the `GithubUpdater`, we have to delete the repo on _repo not found_ case

Reply:
Done!